pages/blog/the-reference-rabbit-hole.md

```diff
@@ -0,0 +1,275 @@
+
+```yaml
+asyncapi: '2.2.0'
+...
+servers:
+  test-server:
+    $ref: './servers/testServer.yaml'
+...
+```
+
+Together with `testServer.yaml`:
+```yaml
+url: ws://mycompany.com/ws
+protocol: ws
+```
+
+Validating the AsyncAPI document using a tool such as [ajv](https://ajv.js.org/) against [JSON Schema representation for 2.2.0](https://github.com/asyncapi/spec-json-schemas/blob/master/schemas/2.2.0.json), it would reject it. 
+
+However, because the parser dereferences first, the document that is being validated is this:
+
+```yaml
+asyncapi: '2.2.0'
+...
+servers:
+  test-server:
+    url: ws://mycompany.com/ws
+    protocol: ws
+...
+```
+
+Checkout [parser-js #405](https://github.com/asyncapi/parser-js/issues/405) for more information.
+
+## What about `$id` keyword
+
+The next thoughts that came to out minds was, one of the key differences between our **Reference Object**, and how `$ref` is resolved in JSON Schema Draft 7, is the [$id keyword](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.2). This allows you to define a URI that is used as a base URI. This means that for example a message such as this:
+
+```yaml
+asyncapi: '2.2.0'
+...
+channels:
+  test/channel:
+    publish:
+      message:
+        schemaFormat: application/schema+json;version=draft-07
+        payload: 
+          $id: https://example.com/schemas/test
+          type: object
+          properties:
+            address: 
+              $ref: "/schemas/address"
+...
+```
+
+This will result in the reference for the `address` property, to be looked up at `https://example.com/schemas/address`, because it uses the Base URI in `$id` from the parent schema (`https://example.com`). 
+
+I tried a little test in the [new Studio](https://studio.asyncapi.com/) (Studio uses the parser, so it could be used for an easy test), [which showed that this was not supported by the parser](https://studio.asyncapi.com/?base64=YXN5bmNhcGk6ICcyLjIuMCcKaW5mbzoKICB0aXRsZTogVGVzdCBvdmVycmlkaW5nIGRlcmVmZXJlbmNlZCBvYmplY3RzIAogIHZlcnNpb246ICcxLjAuMCcKY2hhbm5lbHM6CiAgdGVzdDoKICAgIHB1Ymxpc2g6CiAgICAgIG1lc3NhZ2U6CiAgICAgICAgJHJlZjogJyMvY29tcG9uZW50cy9tZXNzYWdlcy9teU1lc3NhZ2UnCmNvbXBvbmVudHM6CiAgbWVzc2FnZXM6CiAgICBteU1lc3NhZ2U6CiAgICAgIHNjaGVtYUZvcm1hdDogYXBwbGljYXRpb24vc2NoZW1hK2pzb247dmVyc2lvbj1kcmFmdC0wNwogICAgICBuYW1lOiBNeU1lc3NhZ2UKICAgICAgcGF5bG9hZDoKICAgICAgICAgICRpZDogJ2h0dHA6Ly9sb2NhbGhvc3QuY29tLycKICAgICAgICAgIHR5cGU6IG9iamVjdAogICAgICAgICAgcHJvcGVydGllczoKICAgICAgICAgICAgc2VudEF0OgogICAgICAgICAgICAgICRyZWY6ICIvY29tcG9uZW50cy9zY2hlbWFzL3NlbnRBdCI=). The library tries to resolve the reference at `https:///components/schemas/sentAt` when it should have tried to resolve it from `http://localhost.com/components/schemas/sentAt`. See [parser-js #403](https://github.com/asyncapi/parser-js/issues/403) for more information.
```

Comment:
Why dont you use the same example as in 136:146 ?

Reply:
Good question, changing it.